File: plugins/pubsub/ios/Sources/TinyPubSub.swift
//
//  TinyPubSub.swift
//  PlayerUI
//

import Foundation
import JavaScriptCore

import PlayerUI

/**
 A handle to a JavaScript TinyPubSub instance that can be shared across multiple PubSubPlugin instances.

 Create with a `JSContext` for immediate use, or create without one and pass to a `PubSubPlugin`
 which will initialize it automatically during setup. Subscribe and publish calls made before the
 instance is initialized are silently ignored.

 ```swift
 // Standalone usage — provide a context at init:
 let sharedBus = TinyPubSub(context: someContext)

 // Plugin usage — the plugin initializes the shared bus for you:
 let sharedBus = TinyPubSub()
 let plugin1 = PubSubPlugin([], pubsub: sharedBus)
 let plugin2 = PubSubPlugin([], options: PubSubPluginOptions(expressionName: "customPublish"), pubsub: sharedBus)
 ```
 */
public class TinyPubSub {
    /// Reference to the underlying JS TinyPubSub instance. Set during initialization.
    internal var jsValue: JSValue?

    /**
     Creates a TinyPubSub instance and immediately constructs the underlying JS object
     in the given context.
     - parameters:
        - context: The JS context in which the TinyPubSub instance will live
     */
    public init(context: JSContext) {
        initialize(in: context)
    }

    /**
     Creates an uninitialized TinyPubSub handle. Pass this to a `PubSubPlugin` and it will
     initialize the underlying JS instance when the plugin is set up.
     */
    public init() {}

    /**
     Initializes the underlying JS TinyPubSub instance in the given context. Called automatically
     by `PubSubPlugin` during setup. Subsequent calls are no-ops.
     - parameters:
        - context: The JS context in which to create the TinyPubSub instance
     */
    internal func setup(context: JSContext) {
        guard jsValue == nil else { return }
        initialize(in: context)
    }

    private func initialize(in context: JSContext) {
        if let url = ResourceUtilities.urlForFile(name: "PubSubPlugin.native", ext: "js", bundle: Bundle.module),
           let jsString = try? String(contentsOf: url, encoding: .utf8) {
            context.evaluateScript(jsString)
        }
        jsValue = context.evaluateScript("new PubSubPlugin.TinyPubSub()")
    }

    /**
     Subscribe to an event on the shared bus.
     - parameters:
        - eventName: The name of the event to listen for
        - callback: Called with the event name and any associated data when the event fires
     - returns: A subscription token that can be passed to `unsubscribe` to cancel the subscription,
                or `nil` if the JS instance has not been set up yet.
     */
    @discardableResult
    public func subscribe(eventName: String, callback: @escaping (String, AnyType?) -> Void) -> String? {
        guard let ref = jsValue else { return nil }
        let block: @convention(block) (JSValue?, JSValue?) -> Void = { (event, data) in
            guard let name = event?.toString() else { return }
            if
                let isString = data?.isString, isString,
                let objectString = data?.toString()
            {
                callback(name, .string(data: objectString))
            } else if
                let object = data?.toObject(),
                let objectData = try? JSONSerialization.data(withJSONObject: object),
                let eventData = try? AnyTypeDecodingContext(rawData: objectData)
                    .inject(to: JSONDecoder())
                    .decode(AnyType.self, from: objectData)
            {
                callback(name, eventData)
            } else {
                callback(name, nil)
            }
        }
        let jsCallback = JSValue(object: block, in: ref.context) as Any
        return ref.invokeMethod("subscribe", withArguments: [eventName, jsCallback])?.toString()
    }

    /**
     Publish an event on the shared bus.
     - parameters:
        - eventName: The name of the event
        - eventData: Arbitrary data associated with the event
     */
    public func publish(eventName: String, eventData: AnyType) {
        guard
            let ref = jsValue,
            let data = try? JSONEncoder().encode(eventData),
            let dataString = String(data: data, encoding: .utf8),
            let dataObject = ref.context.evaluateScript("(\(dataString))")
        else { return }
        ref.invokeMethod("publish", withArguments: [eventName, dataObject])
    }

    /**
     Unsubscribe using a token returned from a previous `subscribe` call.
     - parameters:
        - token: The token returned by `subscribe`
     */
    public func unsubscribe(token: String) {
        jsValue?.invokeMethod("unsubscribe", withArguments: [token])
    }
}

File: plugins/pubsub/jvm/src/main/kotlin/com/intuit/playerui/plugins/pubsub/TinyPubSub.kt
package com.intuit.playerui.plugins.pubsub

import com.intuit.playerui.core.bridge.Node
import com.intuit.playerui.core.bridge.getInvokable
import com.intuit.playerui.core.bridge.runtime.Runtime

/**
 * A handle to a JavaScript TinyPubSub instance that can be shared across multiple [PubSubPlugin]
 * instances. Create one and pass it to multiple plugin constructors so they all share the same
 * event bus.
 *
 * The underlying JS instance is created the first time a [PubSubPlugin] that holds this handle
 * is applied to a runtime. [subscribe], [publish], and [unsubscribe] calls made before that
 * point will throw.
 *
 * ```kotlin
 * val sharedBus = TinyPubSub()
 * val plugin1 = PubSubPlugin(sharedPubSub = sharedBus)
 * val plugin2 = PubSubPlugin(Config("customPublish"), sharedPubSub = sharedBus)
 *
 * // After the player is set up, use the bus directly:
 * sharedBus.subscribe("myEvent") { event, data -> /* ... */ }
 * ```
 */
public class TinyPubSub {
    internal var node: Node? = null
    private var jsVarName: String? = null

    /**
     * Called by [PubSubPlugin] after the bundle has been loaded into [runtime].
     * Creates the JS TinyPubSub instance on first call; subsequent calls are no-ops and
     * return the same variable name.
     */
    internal fun getOrCreate(runtime: Runtime<*>): String =
        jsVarName ?: run {
            val varName = "_playerui_pubsub_${System.identityHashCode(this)}"
            runtime.execute("globalThis.$varName = new PubSubPlugin.TinyPubSub()")
            node = runtime.execute("globalThis.$varName") as? Node
            jsVarName = varName
            varName
        }
    }

    /**
     * Subscribe to an event
     * @param eventName The name of the event to attach a handler to
     * @param block The handler to call when the event is received
     * @return subscription token used to [unsubscribe]
     */
    public fun subscribe(eventName: String, block: (String, Any?) -> Unit): String =
        node?.getInvokable<String>("subscribe")!!(eventName, block)

    /**
     * Cancel subscription registered with [token]
     * @param token subscription token obtained from [subscribe] call
     */
    public fun unsubscribe(token: String) {
        node?.getInvokable<Any?>("unsubscribe")!!(token)
    }

    /**
     * Publish an event through this pubsub instance
     * @param eventName The name of the event
     * @param eventData Arbitrary data associated with the event
     */
    public fun publish(eventName: String, eventData: Any) {
        node?.getInvokable<Any?>("publish")!!(eventName, eventData)
    }
}
